Flow: add loki.source.syslog component (#2674)
grafana/agent: docs/sources/flow/reference/components/loki.source.syslog.md

`forward_to` | `list(LogsReceiver)`   | List of receivers to send log entries to. | | yes

## Blocks

The following blocks are supported inside the definition of
`loki.source.syslog`:

Hierarchy | Name | Description | Required
--------- | ---- | ----------- | --------
listener | [config][] | Configures a listener for IETF Syslog (RFC5424) messages. | no
listener > tls_config | [tls_config][] | Configures TLS settings for connecting to the endpoint for TCP connections. | no

The `>` symbol indicates deeper levels of nesting. For example, `config > tls_config` 
refers to a `tls_config` block defined inside a `config` block.

[listener]: #listener-block
[tls_config]: #tls_config-block

### listener block

The `listener` block defines the listen address and protocol where the listener
expects syslog messages to be sent to, as well as its behavior when receiving
messages.

The following arguments can be used to configure a `listener`. Only the
`address` field is required and any omitted fields take their default
values.

Name                     | Type          | Description | Default | Required
------------------------ | ------------- | ----------- | ------- | --------

> `address`                | `string`      | The `<host:port>` address to listen to for syslog messages. | | yes

Reviewer: Should be using placeholder syntax here (so people don't think `<>` is something they need to type):

```suggestion
`address`                | `string`      | The `HOST:PORT` address to listen to for syslog messages. | | yes
```

Author: Thanks for the catch!

Multiple `loki.source.syslog` components can be specified by giving them
different labels.

## Usage

```river
loki.source.syslog "LABEL" {
  config {
    listen_address = "LISTEN_ADDRESS"
  }
  ...

  forward_to = RECEIVER_LIST
}
```

## Arguments

`loki.source.syslog` supports the following arguments:

Name         | Type                   | Description          | Default | Required
------------ | ---------------------- | -------------------- | ------- | --------
`forward_to` | `list(LogsReceiver)`   | List of receivers to send log entries to. | | yes

## Blocks

The following blocks are supported inside the definition of
`loki.source.syslog`:

Hierarchy | Name | Description | Required
--------- | ---- | ----------- | --------
listener | [config][] | Configures a listener for IETF Syslog (RFC5424) messages. | no
listener > tls_config | [tls_config][] | Configures TLS settings for connecting to the endpoint for TCP connections. | no

The `>` symbol indicates deeper levels of nesting. For example, `config > tls_config` 
refers to a `tls_config` block defined inside a `config` block.

[listener]: #listener-block
[tls_config]: #tls_config-block

### listener block

> The `listener` block defines the listen address and protocol where the listener
> will expect syslog messages to, as well as its behavior when receiving
> messages.

Reviewer 2: ```suggestion
The `listener` block defines the listen address and protocol where the listener
expects syslog messages to be sent to, as well as its behavior when receiving
messages.
```

Reviewer 2: Something seemed to be missing from this sentence, so I took a guess...

Author: Thanks! And 🤦 about the future tense _once more_, I still haven't gotten used to it.

Reviewer 2: > Thanks! And 🤦 about the future tense _once more_, I still haven't gotten used to it.

Haha, no worries!

The following arguments can be used to configure a `listener`. Only the
`listen_address` field is required and any omitted fields take their default
values.

Name                     | Type          | Description | Default | Required
------------------------ | ------------- | ----------- | ------- | --------

> `address`                | `string`      | The `<host:port>` address to listen on for syslog messages. | | yes

Reviewer 2: ```suggestion
`address`                | `string`      | The `<host:port>` address to listen to for syslog messages. | | yes
```

> `address`                | `string`      | The `<host:port>` address to listen on for syslog messages. | | yes

Reviewer 3: Should this be `listen_address`?

Author: I initially had it this way, but found it was a little stuttering when configuring the top-level block.
```
	listener {
		listen_address  = "127.0.0.1:51898"
		listen_protocol = "udp"
```

Reviewer 3: Fine with that. That aligns with our command line flags which use address.

> `protocol`               | `string`      | The protocol to listen on for syslog messages. Must be either `tcp` or `udp`. | `tcp` | no

Reviewer 2: ```suggestion
`protocol`               | `string`      | The protocol to listen to for syslog messages. Must be either `tcp` or `udp`. | `tcp` | no
```

`idle_timeout`           | `duration`    | The idle timeout for tcp connections. | `"120s"` | no
`label_structured_data`  | `bool`        | Whether to translate syslog structured data to loki labels. | `false` | no
`labels`                 | `map(string)` | The labels to associate with each received syslog record. | `{}` | no
`use_incoming_timestamp` | `bool`        | Whether to set the timestamp to the incoming syslog record timestamp. | `false` | no

> `use_rfc5424_message`    | `bool`        | Whether the full RFC5424-formatted syslog message should be forwarded. | `false` | no

Reviewer 2: ```suggestion
`use_rfc5424_message`    | `bool`        | Whether to forward the full RFC5424-formatted syslog message. | `false` | no
```

`max_message_length`     | `int`         | The maximum limit to the length of syslog messages. | `8192` | no

> By default, the component will assign the log entry timestamp as the time it
> was processed.

Reviewer 2: ```suggestion
By default, the component assigns the log entry timestamp as the time it
was processed.
```

> The `labels` map will be applied to every message that the component reads.

Reviewer 2: ```suggestion
The `labels` map is applied to every message that the component reads.
```

> All header fields from the parsed RFC5424 messages will be brought in as
> internal labels, prefixed with `__syslog_`.

Reviewer 2: ```suggestion
All header fields from the parsed RFC5424 messages are brought in as
internal labels, prefixed with `__syslog_`.
```

> If `label_structured_data` is set, structured data in the syslog header will
> also be translated to internal labels in the form of
> `__syslog_message_sd_<ID>_<KEY>`. A structured data entry of `[example@99999
> test="yes"]` would become the label `__syslog_message_sd_example_99999_test`
> with the value `"yes"`.

Reviewer 2: ```suggestion
If `label_structured_data` is set, structured data in the syslog header are
also translated to internal labels in the form of
`__syslog_message_sd_<ID>_<KEY>`. For example, a structured data entry of `[example@99999
test="yes"]` becomes the label `__syslog_message_sd_example_99999_test`
with the value `"yes"`.
```

### tls_config block

{{< docs/shared lookup="flow/reference/components/tls-config-block.md" source="agent" >}}

## Exported fields

`loki.source.syslog` does not export any fields.

## Component health

`loki.source.syslog` is only reported as unhealthy if given an invalid
configuration.

## Debug information

`loki.source.syslog` exposes some debug information per syslog listener:
* Whether the listener is currently running.
* The listen address.
* The labels that the listener applies to incoming log entries.

## Debug metrics
* `loki_source_syslog_entries_total` (counter): Total number of successful entries sent to the syslog component.
* `loki_source_syslog_parsing_errors_total` (counter): Total number of parsing errors while receiving syslog messages.
* `loki_source_syslog_empty_messages_total` (counter): Total number of empty messages received from the syslog component.

## Example

> This example collects listens for Syslog messages in valid RFC5424 format over
> TCP and UDP in the specified ports and forwards them to a `loki.write`
> component.

Reviewer 2: ```suggestion
This example listens for Syslog messages in valid RFC5424 format over
TCP and UDP in the specified ports and forwards them to a `loki.write`
component.
```
